src/theory/datatypes/inference_manager.cpp

```diff
@@ -0,0 +1,190 @@
+/*********************                                                        */
+/*! \file inference_manager.cpp
+ ** \verbatim
+ ** Top contributors (to current version):
+ **   Andrew Reynolds
+ ** This file is part of the CVC4 project.
+ ** Copyright (c) 2009-2020 by the authors listed in the file AUTHORS
+ ** in the top-level source directory) and their institutional affiliations.
+ ** All rights reserved.  See the file COPYING in the top-level source
+ ** directory for licensing information.\endverbatim
+ **
+ ** \brief Datatypes inference manager
+ **/
+
+#include "theory/datatypes/inference_manager.h"
+
+#include "expr/dtype.h"
+#include "options/datatypes_options.h"
+#include "theory/theory.h"
+
+using namespace CVC4::kind;
+
+namespace CVC4 {
+namespace theory {
+namespace datatypes {
+
+InferenceManager::InferenceManager(Theory& t,
+                                   TheoryState& state,
+                                   ProofNodeManager* pnm)
+    : InferenceManagerBuffered(t, state, pnm),
+      d_lemmasSent(t.getSatContext()),
+      d_addedLemma(false),
+      d_addedFact(false)
+{
+  d_true = NodeManager::currentNM()->mkConst(true);
+}
+
+void InferenceManager::reset()
+{
+  d_addedLemma = false;
+  d_addedFact = false;
+}
+
+bool InferenceManager::mustCommunicateFact(Node n, Node exp) const
```

Comment:
I am wondering if this should be a general method in the `InferenceManager` class that the individual theories override?

Reply:
It's possible as a next step.  The even more pressing candidate is actually the d_hasAddedFact / d_hasAddLemma / reset functionality.  That is used in many theories.